templateargs/process.go

```diff
@@ -0,0 +1,103 @@
+package templateargs
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/yarpc/yab/templateargs/argparser"
+
+	"gopkg.in/yaml.v2"
+)
+
+// ProcessMap takes a YAML request that may contain values like ${name:prashant}
+// and replaces any template arguments with those specified in args.
+func ProcessMap(req map[interface{}]interface{}, args map[string]string) (map[interface{}]interface{}, error) {
+	argsUnmarshalled := make(map[string]interface{}, len(args))
+
+	for k, v := range args {
+		var unmarshalled interface{}
+		if err := yaml.Unmarshal([]byte(v), &unmarshalled); err != nil {
+			return nil, fmt.Errorf("failed to process %q with value %q: %v", k, v, err)
+		}
+		argsUnmarshalled[k] = unmarshalled
+	}
+
+	return processMap(req, argsUnmarshalled)
+}
+
+func processString(v string, args map[string]interface{}) (interface{}, error) {
```

Comment:
This function should unconditionally parse the string in its DSL, including escaping `\$` and `\\`. The grammar should return a simple AST with either a Lit or a Var term, and this function should just switch on that.

This leaves open the possibility of extending the grammar to support string interpolation like `${host}:${port}`. We can forbid this for now.